Additional tests for `ios_get_app_version`

iangmaia opened the pull request:
**EDIT**: This PR was re-created in the main repo as https://github.com/wordpress-mobile/release-toolkit/pull/451

## What does it do?
This PR improves the existing method `Ios::VersionHelper::read_from_config_file()`, making it more reliable and robust by using the gem `xcodeproj`, as well as adding better error handling. Some additional tests cases for the code were also implemented.

This need came up as I was testing the code and noticed that possible format variations for the `.xcconfig` file were not being treated properly and there was no error handling.

## Related PRs
- This PR builds on top of [#445](https://github.com/wordpress-mobile/release-toolkit/pull/445)

## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the approprioate existing `###` subsection of the existing `## Trunk` section.

iangmaia committed:
Improve the version parsing code to better handle error cases

iangmaia committed:
Add unit tests for ios_get_app_version error cases

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_version_helper.rb:
```diff
@@ -306,15 +309,22 @@ def self.read_build_number_from_config_file(filePath)
-          File.open(filePath, 'r') do |f|
+        def self.read_from_config_file(key, file_path)
+          UI.user_error!("File #{file_path} not found") unless File.exist?(file_path)
+
+          File.open(file_path, 'r') do |f|
             f.each_line do |line|
-              line = line.strip()
-              return line.split('=')[1] if line.start_with?("#{key}=")
+              line.strip!
+              next if line.nil? || line.empty?
+
+              key_value = line.split(/\s*=\s*/)
+              if key_value[0].strip() == key
+                return key_value[1].strip()
+              end
```
Nice improvement 👍 

What do you think about using the `xcodeproj` gem to [read the `xcconfig` file](https://rubydoc.info/gems/xcodeproj/Xcodeproj/Config) instead though? Would probably make this even more resilient, given I trust `Xcodeproj::Config` is [already well tested](https://github.com/CocoaPods/Xcodeproj/blob/master/spec/config_spec.rb#L248-L269) on those edge cases 🙂 

PS: [That `xcodeproj` gem is already a dependency of `fastlane`](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/Gemfile.lock#L175-L211) so it's not like adding it as an explicit dependency of the `release-toolkit` plugin would add any dependency or extra download to our client projects anyway, as they would already have it.

iangmaia replied:
Ah, yes, in principle it didn't occur to me, but when I read your comment in the base PR I also thought this would be much better indeed 👍

iangmaia replied:
Implemented here https://github.com/iangmaia/release-toolkit/pull/1/commits/02a99f3f822de360db6587ce97ecf19fb25b2f0e

iangmaia committed:
Make ios_get_app_version_spec unit tests to work with an actual file to avoid relying on implementation details

iangmaia committed:
Improve method to read .xcconfig file to use Xcodeproj::Config instead of manually parsing it

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_version_helper.rb:
```diff
@@ -1,3 +1,5 @@
+require 'xcodeproj'
```
We need to make sure this gem is listed in the toolkit's gemspec.

Even if in practice, because the release-toolkit plugin will only make sense when used with fastlane, and fastlane already has xcodeproj as a transitive dependency, imho it still makes sense to make it explicit that we our tookit gem on the that xcodeproj gem.

iangmaia replied:
Agreed 👍

iangmaia replied:
Added here https://github.com/iangmaia/release-toolkit/pull/1/commits/09ade61cdf715da1f381e56d93c2ecc4f2e092c6

AliSoftware commented on spec/ios_get_app_version_spec.rb:
```diff
@@ -5,37 +5,54 @@
+        VERSION_SHORT = 6
         // a comment
-        VERSION_LONG=6.30.1
+        VERSION_LONG = 6.30.1
```
Interestingly, that means that now we do have tests for the cases where there are spaces around the key and value… but we also don't have tests for cases without spaces around the `=` anymore. Could it be worth duplicating those 2 tests but have one with spaces and one without?

iangmaia replied:
My idea here was that the spaces would be where parsing would fail. But you're right, it's probably worth it adding this test as well as some sort of invalid format.

iangmaia replied:
Implemented here: https://github.com/iangmaia/release-toolkit/pull/1/commits/d72c9366bc8d28da8113808ca06579885f90fe91

mokagio commented on spec/ios_get_app_version_spec.rb:
```diff
@@ -5,37 +5,54 @@
+    end
+
+    def expect_version(xcconfig_mock_content:, expected_version:)
+      allow(File).to receive(:exist?).and_return(true)
```
Nitpick. Written like this, the `allow` seems too "generous". It will pass for any input, which I think could backfire on us.

Also, do we even need it now that 

1. There is a dedicated test for "throws an error when the file is not found" which doesn't use this helper
2. This helper creates a file, making the stubbing of the `File` call redundant

?

iangmaia replied:
Oh indeed, you're right on the too lenient `allow`, it should have had the parameter... but as you point out, as a file is being created and passed, this isn't necessary anymore. Will change it, nice catch!

iangmaia replied:
`allow` removed here: https://github.com/iangmaia/release-toolkit/pull/1/commits/f64166736ab1db4b5f798ae16b531a0159205d3b

iangmaia committed:
Add explicit dependency to xcodeproj

iangmaia committed:
Remove unnecessary mock of File existence check

iangmaia committed:
Add more test cases for different .xcconfig file formats

iangmaia commented:
Hi, @AliSoftware and @mokagio! I've addressed the remaining comments.
If everything is ok, we could merge https://github.com/wordpress-mobile/release-toolkit/pull/445 and immediately after I'll re-create this PR targeting the main repo, like I did for the `skip_glotpress` PR.

By the way, just because I never faced such situation working in different forks, I was thinking of a couple of alternatives in terms of PR strategy for such cases (having PRs based on other PRs from different forks), other than the strategy I followed:
1. Once approved, merge the "stacked" PR in my repo directly in the base branch, with the disadvantage of having a mixed PR final merge to the main repo, containing both (or more) PRs.
2. Create the "stacked" PR  directly in the main repo as a draft, with the disadvantage of this draft containing the original branch diff until the original branch is merged.
3. Wait until the base branch is merged to create the follow up PR, with the disadvantage of making the merges dependent and delaying reviews that could be done in parallel, potentially also by different people.

All in all I think having the PRs in different repos like I did, even with some extra admin overhead, it sounds like a good compromise considering that the reviews can be done with a smaller diff and in parallel, to me being the main point of doing "stacked" PRs like this. Also, the final git history remains clean with separate merge commits on the main repo's trunk.

AliSoftware commented on fastlane-plugin-wpmreleasetoolkit.gemspec:
```diff
@@ -37,6 +37,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'progress_bar', '~> 1.3'
   spec.add_dependency 'rake', '>= 12.3', '< 14.0'
   spec.add_dependency 'rake-compiler', '~> 1.0'
+  spec.add_dependency 'xcodeproj', '~> 1.22.0'
```
This is too specific of a constraint.

That `~>` means that it will only allow to blindly update the last component of the version, so `~>1.22.0` will only allow us to update to any `1.22.z` with `z>=0`. But semantic versioning makes it ok to also get any new minor version — so as long as it's still a major version 1 (`1.y.z`) we should not refrain xcodeproj from being updated.

This is particularly important to have a more relaxed constraint here because `xcodeproj` is also used by other gems of our client apps' dependency tree, especially its a dependency of both `cocoapods` and `fastlane`, and we don't want the release-toolkit to generate a dependency conflict and prevent `xcodeproj` to be updated to a newer non-breaking / minor version is one of those gems require it in the future.

iangmaia replied:
True, minor updates should be ok in general 👍 besides that, I didn't immediately realise the issue with client apps' dependency tree (to be fair, even release-toolkit own's dependency tree I guess) in this case, good point.

iangmaia replied:
Done in 70e3dc38421823ee71c7f7c55d53c734b7b4bfc7

AliSoftware commented:
> If everything is ok, we could merge wordpress-mobile#445 and immediately after I'll re-create this PR targeting the main repo, like I did for the `skip_glotpress` PR

While I'm good for this in theory, in practice since 445 is a breaking change that will _actually_ require small changes to our client apps before we can adopt it (while the previous one about `skip_GlotPress` was technically breaking, but not really in practice because all our client apps have already migrated away from `download_metadata.swift` to the new way)… but it happens that I need to land a couple of PRs of mine in the release toolkit + do a new release very soon to start using some of the new stuff I am working on in client repos ASAP. So the timing wouldn't be ideal for any big breaking change like 445 to land just before, as that would postpone me from moving forward with my current project 😅

Given this what I'm hoping to do is land the final PR I recently made in release toolkit that I need for my work (it has been approved with suggestions already so shouldn't take long to finish up 🤞), release a new version of the toolkit, and then we can merge your PR 445 and move forward like you suggest. All while you continue preparing the `Migration.md` file to document the changes needed by client apps to adopt the next major release once we're ready to ship it.

That might mean that we won't be able to merge your 445 before, say, Tuesday (once I've landed my non-breaking PRs and released a new intermediate version). So slight delay for you, sorry about that 🙃. But I'm sure you have plenty of other things to work on (including Migration doc and bonus task) in the meantime 🙂

iangmaia committed:
Fix dependency to xcodeproj to allow for minor updates

iangmaia commented:
> Given this what I'm hoping to do is land the final PR I recently made in release toolkit that I need for my work (it has been approved with suggestions already so shouldn't take long to finish up 🤞), release a new version of the toolkit, and then we can merge your PR 445 and move forward like you suggest. All while you continue preparing the `Migration.md` file to document the changes needed by client apps to adopt the next major release once we're ready to ship it.
> 
> That might mean that we won't be able to merge your 445 before, say, Tuesday (once I've landed my non-breaking PRs and released a new intermediate version). So slight delay for you, sorry about that 🙃. But I'm sure you have plenty of other things to work on (including Migration doc and bonus task) in the meantime 🙂

Exactly, no worries at all, I had the impression this was going to be the case based on your comments on my last update 🙂

AliSoftware commented:
https://github.com/wordpress-mobile/release-toolkit/pull/445 got merged, so this is ready for a rebase 🙂
